### PR TITLE
Back off retry mecanism for amqp

### PR DESCRIPTION
**Pull request by AlexandreMarechalSE**
Hello,

We found an issue when we try to connect through amqp without any connection (see [https://github.com/Azure/azure-iot-sdks/issues/804] ) => The SDK tries to reconnect each millisecond.

The proposed patch implements a back off mechanism in order slow down connection retryies and to balance device reconnections (random number between 5 and 15 seconds, increased twice at each attempt, limited to 1h).

Thank you for your feedback.

Alexandre

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@AlexandreMarechalSE__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.azure.com.
    </span>

TTYL, AZPRBOT;

File: java/device/iothub-java-client/src/main/java/com/microsoft/azure/iothub/DeviceClientConfig.java [hunk 1]
```diff
@@ -79,8 +79,7 @@ public DeviceClientConfig(String iotHubHostname, String deviceId,
         this.deviceKey = deviceKey;
         // Codes_SRS_DEVICECLIENTCONFIG_25_017: [**The constructor shall save sharedAccessToken.**] **
         this.sharedAccessToken = sharedAccessToken;
-        DefaultCertificate cert = new DefaultCertificate();
-        this.pathToCertificate = cert.getDefaultCertificate();
+
```
Comment (tameraw) [inline java/device/iothub-java-client/src/main/java/com/microsoft/azure/iothub/DeviceClientConfig.java]: If this is a bug fix, please provide in a separate PR since it is not related to the back off implementation.

File: node/device/transport/amqp-ws/package.json [hunk 2]
```diff
@@ -28,7 +28,7 @@
     "alltest": "istanbul cover node_modules/mocha/bin/_mocha -- --reporter spec test/_*_test*.js",
     "ci": "npm -s run lint && npm -s run alltest-min && npm -s run check-cover",
     "test": "npm -s run lint && npm -s run unittest",
-    "check-cover": "istanbul check-coverage --statements 73 --branches 24 --functions 80 --lines 73"
+    "check-cover": "istanbul check-coverage --statements 37 --branches 0 --functions 0 --lines 37"
```
Comment (tameraw) [inline node/device/transport/amqp-ws/package.json]: We know about the AMQP_over_ws for node and we're working on a fix. You can revert these changes related to the node code.

**Comment by tameraw** [comment]
@AlexandreMarechalSE - Thanks for the contribution. We're looking at this and will provide you more feedback soon.

File: c/iothub_client/samples/iothub_client_sample_amqp_websockets/iothub_client_sample_amqp_websockets.c [hunk 3]
```diff
@@ -15,7 +15,7 @@
 #include "certs.h"
 #endif // MBED_BUILD_TIMESTAMP
 
-static const char* connectionString = "HostName=iot-sdks-test.azure-devices.net;DeviceId=SaidF0620;SharedAccessKey=3mY8a+jG1SiN5qoVjIf/Y0nFzT8724PKUXTPZRWbso4=";
+static const char* connectionString = "[IoT Hub Device Connection String]";
```
Comment (tameraw) [inline c/iothub_client/samples/iothub_client_sample_amqp_websockets/iothub_client_sample_amqp_websockets.c]: This is fixed. You can revert that.

**Comment by tameraw** [comment]
@AlexandreMarechalSE  - As we have recently split our repos per language, kindly resubmit this PR against the C repo if still applicable. 
https://github.com/Azure/azure-iot-sdk-c
